File: packages/python/src/authdog/redirects.py
"""Open-redirect protection for logout / post-auth redirects."""

from __future__ import annotations

import re

__all__ = ["sanitize_redirect_path"]

_SCHEME_RE = re.compile(r"^[a-z][a-z0-9+.-]*:", re.IGNORECASE)
_CONTROL_OR_BACKSLASH_RE = re.compile(r"[\\\x00-\x1f]")


def sanitize_redirect_path(target: object, fallback: str = "/") -> str:
    """Return a safe, same-origin redirect target.

    Only relative paths are allowed. A value is rejected (falling back to
    ``fallback``) if it is empty / not a string, is protocol-relative
    (``//`` or ``/\\``), contains a scheme, or contains a backslash / control
    character that a browser might normalize to ``/``.
    """
    if not isinstance(target, str) or not target:
        return fallback

    if (
        not target.startswith("/")
        or target.startswith("//")
        or target.startswith("/\\")
    ):
        return fallback

    if _CONTROL_OR_BACKSLASH_RE.search(target):
        return fallback

    if _SCHEME_RE.match(target):
        return fallback

    return target

File: packages/go/redirects.go
package authdog

import (
	"regexp"
	"strings"
)

var (
	schemeRe           = regexp.MustCompile(`(?i)^[a-z][a-z0-9+.-]*:`)
	controlOrBackslash = regexp.MustCompile(`[\\\x00-\x1f]`)
)

// SanitizeRedirectPath returns a safe, same-origin redirect target, falling
// back to fallback (use "/" for the default) for anything that could be an
// open redirect.
//
// Only relative paths are allowed. A value is rejected if it is empty, is
// protocol-relative ("//" or "/\"), contains a scheme, or contains a
// backslash / control character a browser might normalize to '/'.
func SanitizeRedirectPath(target, fallback string) string {
	if target == "" {
		return fallback
	}
	if !strings.HasPrefix(target, "/") ||
		strings.HasPrefix(target, "//") ||
		strings.HasPrefix(target, "/\\") {
		return fallback
	}
	if controlOrBackslash.MatchString(target) {
		return fallback
	}
	if schemeRe.MatchString(target) {
		return fallback
	}
	return target
}
